### PR TITLE
Starlark: intern dot-expression RHS identifiers globally

### PR DESCRIPTION
**Pull request by stepancheg**
Indentifiers are interned in two places:
* when parsing dot-expression RHS
* in `CallUtils` map keys

Thus when evaluating dot expression (`foo.bar`), `bar` is interned
and lookup is faster: `String.equals` performs the identity comparison
before falling back to the comparison of contents.

This PR does interning conservatively: it only interns identifiers
from RHS of dot expression, but no other identifiers.

Other identifiers, and even string literals could be interned to
get performance gain. Consider this example:

```
def my_cc_library_wrapper(**kwargs):
  internal = kwargs.pop("is_internal_library", False)
  if internal:
    local_defs = kwargs.pop("local_defines", []) + ["-DINTERNAL"]
    kwargs["local_defines"] = local_defs
  cc_library(**kwargs)

my_cc_library_wrapper(..., is_internal_library = True)
```

Here we lookup by literal "is_internal_library" which is specified
as a named argument "is_internal_library" in another file: if both
strings were interned, lookup would be faster.

But interning everything may have dire performance implications.
Can be done later.

From my experiments running profiler, attribute lookup is one of
significant bottlenecks of the Starlark interpreter.

Benchmark:

```
def test():
    for _ in range(50000000):
        "abc".startswith
        [].append
        "cde".endswith
        [].extend
        {}.pop
        "fg".rsplit
        [].pop

test()
```

Output is:

```
A: N=27 r=13.546+-0.785 se=0.153
B: N=27 r=11.838+-0.965 se=0.189
B/A: 0.874
```

Or with include benchmark (best of 10 runs 10 seconds each):

```
benchmark                   ops     cpu/op    wall/op   steps/op   alloc/op
# before
bench_dot_expr_native   33554431      353ns      351ns         30       529B
# after
bench_dot_expr_native   67108863      298ns      297ns         30       288B
```

### CODE REVIEW

**Review by alandonovan — CHANGES_REQUESTED** [review]
I hope to submit the compiler next week, which will make this change unnecessary.

**Comment by aiuto** [comment]
Is this still relevent?  That is, did the compiler change make it obsolete?

**Comment by alandonovan** [comment]
The compiler work is still in progress. (I have implemented three different approaches to the compiler and runtime and, demoralizingly, none is yet sufficiently faster than the baseline to justify its complexity.)

**Comment by stepancheg** [comment]
> none is yet sufficiently faster than the baseline to justify its complexity

What is sufficiently faster by your definition?

My last attempt to implement bytecode was 20% improvement IIRC. That would be enough for us, because it would make the interpreter faster than Python for our use case.

Additionally,
* enabling bytecode interpreter may unlock certain optimizations which might be inappropriate for doing on AST. Like for example, faster Starlark function calls by performing "linking" caller arguments to parameters (`foo(bar=baz, qux=quux)` can be replaced with `foo(baz, quux)` for global functions)
* there are still several missing optimizations in current implementation, for example, indexed predeclared and universe variables, implementing them will make bytecode interpreter speedup better

**Comment by stepancheg** [comment]
Added another commit on top: inline function parameter names and call argument names to speed up code with many calls with named arguments.

```
A: n=16 mean=9.384 std=0.553 se=0.142 min=8.471 med=9.313
B: n=16 mean=8.903 std=0.559 se=0.144 min=8.148 med=8.906
B/A: 0.949 0.905..0.995 (95% conf)
```

```
def foo(a, bb, c):
  pass

def test():
  for i in range(10):
    print(i)
    for j in range(5000000):
      foo(a=10, bb="rr", c=[])

test()
```

File: src/main/java/net/starlark/java/syntax/StarlarkStringInterner.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+public class StarlarkStringInterner {
+  private StarlarkStringInterner() {}
+
+  private static final Interner<String> INTERNER = Interners.newWeakInterner();
+
+  /** Weak intern the string. */
+  public static String intern(String string) {
+    return INTERNER.intern(string);
+  }
+}
```
Comment (tetromino) [inline src/main/java/net/starlark/java/syntax/StarlarkStringInterner.java:21]: This is going to be the third string interner in Starlark; we already have Lexer.identInterner and Parser.stringInterner.

I would much prefer to coalesce these interners into one somehow, and if that's not possible, at least explain which interner should be used in which circumstance and why.

**Review by tetromino — CHANGES_REQUESTED** [review]

**Comment by brandjon** [comment]
Note that string interning was changed by 9bf5a14850685c5b0c6231a47809701528d58af1 to intern all idents together, so that'll affect the before/after benchmarks if this change is rebased on head. For instance, that should mean that param names and keyword arg names are already interned together.

There may still be some advantage in further unifying interning across the Starlark interpreter, so ident interning merges with string literal interning (which is currently in the parser since it handles the `'a' + 'b'` -> `'ab'` static optimization) and also with CallUtils' method name lookup. I think in 9bf5a14850685c5b0c6231a47809701528d58af1 I simply considered that kind of experimentation out of scope.

**Pull request closed without merging**